### ISSUE TITLE
DoclingBERT v3: Spatial-Aware 7-Class Architecture

### ISSUE THREAD

**Issue by donaldbraman** [donaldbraman/docling-testing#18]
# DoclingBERT v3: Spatial-Aware 7-Class Architecture

## Problem Statement

DoclingBERT v2 achieved 84% F1 on body_text classification but has critical limitations:
- **Only uses 3 of 7 available classes** (body_text, footnote, caption)
- **Text-only features** - cannot distinguish position-dependent elements
- **Missing 4 classes** with training data: heading, page_header, page_footer, cover

Legal documents have highly structured layouts where **spatial position is as important as text content**:
- Headers/footers: Always at top/bottom margins
- Footnotes: Bottom of page, smaller text
- Headings: Distinct positioning and size
- Cover pages: Unique layout patterns

## Solution: Full Spatial Embeddings Architecture

Implement state-of-the-art spatial document understanding following LayoutLMv3/Spatial ModernBERT:

### Technical Approach

**1. Spatial Feature Integration**
- Extract bounding box coordinates from existing Docling JSON files (179 PDFs)
- Normalize bbox to [0-999] range: x0, y0, x1, y1, width, height
- Create 6 dedicated embedding layers (1000 vocab × 128 dim each)
- Fuse spatial + text embeddings: 6 × 128 = 768 dimensions

**2. Architecture**
```python
text_embed (768) + spatial_embed (768) → ModernBERT → Classifier (7 classes)
```

**3. Training Strategy**
- Focal Loss (α=0.25, γ=2.0) for class imbalance
- Class weighting for minority classes
- 1000 training steps with eval every 50 steps
- Early stopping on validation loss

**4. 7-Class Schema**
1. body_text
2. heading
3. footnote
4. caption
5. page_header
6. page_footer
7. cover

## Research Foundation

**Proven approach from:**
- LayoutLM (Microsoft Research, 2019): 0-1000 normalization standard
- LayoutLMv2 (2021): Spatial embeddings architecture
- LayoutLMv3 (2022): 95%+ F1 on document understanding
- Spatial ModernBERT (2024): 96.91% TEDS on table extraction

### Phase 1: Corpus Preparation (1-2 days)
- [ ] Parse 179 Docling JSON files for bbox coordinates
- [ ] Normalize bbox to [0-999] range
- [ ] Create balanced 7-class corpus (~12,000-15,000 samples)
- [ ] Output: `data/spatial_7class_corpus.csv` with bbox columns

### Phase 2: Model Architecture (2-3 days)
- [ ] Implement `SpatialModernBERT` with 6 embedding layers
- [ ] Create custom DataLoader for bbox tensors
- [ ] Implement Focal Loss
- [ ] Add comprehensive logging/monitoring

### Phase 3: Training (3-4 days)
- [ ] Train with spatial embeddings (1000 steps)
- [ ] Evaluate every 50 steps
- [ ] Generate confusion matrices
- [ ] Track per-class metrics
- [ ] Hyperparameter tuning if needed

### Phase 4: Evaluation (1-2 days)
- [ ] Comprehensive metrics (per-class F1, confusion matrix)
- [ ] Ablation study (text-only vs spatial)
- [ ] Qualitative error analysis (100 misclassified samples)
- [ ] Production readiness validation

## Deliverables

**Code:**
- `scripts/corpus_building/build_spatial_corpus.py`
- `scripts/training/spatial_modernbert.py`
- `scripts/training/train_spatial_v3.py`
- `scripts/evaluation/evaluate_spatial_v3.py`

**Model:**
- `models/doclingbert-v3-spatial/final_model/`
- `label_map.json` with full 7-class schema

**Documentation:**
- `docs/reports/doclingbert_v3_training_report.md`
- `docs/reports/spatial_ablation_study.md`

**Tests:**
- `tests/test_spatial_corpus.py`
- `tests/test_spatial_model.py`
- `tests/test_spatial_inference.py`

## Success Criteria

**Minimum Viable:**
- ✅ body_text F1 ≥ 88% (vs v2: 84%)
- ✅ All 7 classes F1 ≥ 75%
- ✅ Spatial features provide ≥ 5% improvement

**Target Performance:**
- ✅ body_text F1 ≥ 92%
- ✅ All 7 classes F1 ≥ 85%
- ✅ Balanced accuracy ≥ 90%

**Production Ready:**
- ✅ Inference < 2 sec/document
- ✅ Stable across PDF formats
- ✅ Interpretable error patterns

## Timeline

**Total: 7-11 days**
- Days 1-2: Corpus extraction with bbox
- Days 3-5: Architecture implementation
- Days 6-9: Training + tuning
- Days 10-11: Evaluation + iteration

## Version

This will be **v3.0.0** (major version bump for architectural change)

## References

- [LayoutLMv3 Paper](https://arxiv.org/abs/2204.08387)
- [Spatial ModernBERT](https://arxiv.org/html/2507.08865)
- [Multiclass Training Strategy](docs/multiclass-training.md)

---

**Labels:** enhancement, ml-model, high-priority
